### ISSUE TITLE
NaNs when many of the logits are masked (-inf)

### ISSUE THREAD

**Issue by YonatanGideoni** [Dao-AILab/flash-attention#1930]
When many of the logits are -inf, as may happen when using some kinds of masking (e.g. top-p sampling), flash-attn's `cross_entropy_loss` can give NaNs:

```
import torch
from flash_attn.ops.triton.cross_entropy import cross_entropy_loss

V = 100000
row = torch.full((V,), float('-inf'), device='cuda', dtype=torch.float32)
i = 50000
row[i] = 10
logits = row.unsqueeze(0)  # [1, V]
labels = torch.tensor([i], device='cuda', dtype=torch.long)

loss = cross_entropy_loss(logits, labels)[0]
print(torch.isfinite(loss))
print(loss)
```

Likely occurs due to how the tiling is calculated internally, as for some values of `i` (like 0) this doesn't happen.

Versions:
Flash-attn - 2.8.3
torch - 2.8.0
cuda - 12.5
Python - 3.10.18

**Comment by tongyx361** [comment]
It seems that this occurs when a whole block of logits are `-inf`.
Would it be feasible to mask with some finite value?

```python
import torch
from flash_attn.ops.triton.cross_entropy import cross_entropy_loss

V = 32768
stride = 8192
for mask_value in [float('-inf'), -100, -50]:
    for i in range(0, V, stride):
        row = torch.full((V,), mask_value, device='cuda', dtype=torch.float32)
        row[i] = 10
        logits = row.unsqueeze(0)  # [1, V]
        labels = torch.tensor([i], device='cuda', dtype=torch.long)
        
        loss = cross_entropy_loss(logits, labels)[0]
        print(f"{mask_value=}, {i=}, {loss=}")
        if not torch.isfinite(loss):
            print("Infinite loss!")
            break
```

```
mask_value=-inf, i=0, loss=tensor([0.], device='cuda:0')
mask_value=-inf, i=8192, loss=tensor([0.], device='cuda:0')
mask_value=-inf, i=16384, loss=tensor([nan], device='cuda:0')
Infinite loss!
mask_value=-100, i=0, loss=tensor([0.], device='cuda:0')
mask_value=-100, i=8192, loss=tensor([0.], device='cuda:0')
mask_value=-100, i=16384, loss=tensor([0.], device='cuda:0')
mask_value=-100, i=24576, loss=tensor([0.], device='cuda:0')
mask_value=-50, i=0, loss=tensor([0.], device='cuda:0')
mask_value=-50, i=8192, loss=tensor([0.], device='cuda:0')
mask_value=-50, i=16384, loss=tensor([0.], device='cuda:0')
mask_value=-50, i=24576, loss=tensor([0.], device='cuda:0')
```

Versions:

```
flash-attn=2.7.4.post1
```

`16384` corresponds with `MAX_BLOCK_SIZE = 16 * 1024`

https://github.com/Dao-AILab/flash-attention/blob/5231d95fe13733fb534c01895f7ea88c6a6c7793/flash_attn/ops/triton/cross_entropy.py#L179-L180

The first `nan` might appear when `m_i_new` is `-inf`:

https://github.com/Dao-AILab/flash-attention/blob/5231d95fe13733fb534c01895f7ea88c6a6c7793/flash_attn/ops/triton/cross_entropy.py#L59-L60
